hedera-node/hedera-app/src/main/java/com/hedera/node/app/services/ServiceProviderImpl.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.services;
+
+import com.hedera.node.app.spi.FacilityFacade;
+import com.hedera.node.app.spi.service.Service;
+import com.hedera.node.app.spi.service.ServiceFactory;
+import com.hedera.node.app.spi.service.ServiceProvider;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.ServiceLoader.Provider;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ServiceProviderImpl implements ServiceProvider {
+
+    private final Map<Class<Service>, Service> services;
+
+    public ServiceProviderImpl(final FacilityFacade facilityFacade) {
+        services = new HashMap<>();
+        final ServiceLoader<ServiceFactory> serviceLoader = ServiceLoader.load(ServiceFactory.class);
+        final Set<ServiceFactory> servicesFactories =
+                serviceLoader.stream().map(Provider::get).collect(Collectors.toSet());
+
+        servicesFactories.stream().sorted((a, b) -> sort(a, b)).forEach(factory -> {
```

Comment:
This is a fairly complex cascaded lambda expression.  It might be clearer to use one or more static inner classes (implementing Function and/or BiFunction) to encapsulate this block of code so it's easier to follow.